LiteCore/Database/DatabaseImpl.cc

```diff
@@ -642,30 +659,46 @@ namespace litecore {
 
         // Instantiate it, creating the KeyStore on-disk if necessary:
         KeyStore &store = _dataFile->getKeyStore(keyStoreName);
-        auto collection = make_retained<CollectionImpl>(this, spec, store);
-        // Update its state & add it to _collections:
-        auto collectionPtr = collection.get();
-        _collections.insert({CollectionSpec(collection->getSpec()),
-                             move(collection)});
-        if (isInTransaction())
-            collectionPtr->transactionBegan();
+        if(!collectionPtr) {
+            // This will be entered if the collection never existed or was removed, but
+            // skipped if there is already a collection that was previously invalidated 
+            // (the latter case already reset valid to true near the top of this method)
+            auto collection = make_retained<CollectionImpl>(this, spec, store);
+            // Update its state & add it to _collections:
+            collectionPtr = collection.get();
+            _collections.insert({CollectionSpec(collection->getSpec()),
+                                 move(collection)});
+
+            if (isInTransaction())
+                collectionPtr->transactionBegan();
+        }
+
+        
+
+        collectionPtr->setValid(true);
         return collectionPtr;                                               //-> New object
     }
 
 
     void DatabaseImpl::deleteCollection(CollectionSpec spec) {
-        Transaction t(this);
+        {
+            Transaction t(this);
 
-        LOCK(_collectionsMutex);
-        if (auto i = _collections.find(spec); i != _collections.end()) {
-            asInternal(i->second.get())->close();
-            _collections.erase(i);
+            LOCK(_collectionsMutex);
+            if (auto i = _collections.find(spec); i != _collections.end()) {
+                asInternal(i->second.get())->close();
+                _collections.erase(i);
+            }
+            _dataFile->deleteKeyStore(collectionNameToKeyStoreName(spec));
+            if (isDefaultCollection(spec))
+                _defaultCollection = nullptr;
+
+            t.commit(); 
         }
-        _dataFile->deleteKeyStore(collectionNameToKeyStoreName(spec));
-        if (isDefaultCollection(spec))
-            _defaultCollection = nullptr;
 
-        t.commit();
+        _dataFile->forOtherDataFiles([spec](DataFile* df) {
```

Comment:
After today's meeting I feel more strongly to move the above statement inside the transaction.

Reply:
I think I see the problem you are trying to point out, but a transaction is not going to help.

The problem would be if a collection is deleted on one thread and created on another in between the collection invalidation and the datafile reset.  The creation would not recreate the table because of the existence flag.  A transaction is not going to stop this logic though because it has nothing to do with the database.  A transaction would force the CREATE TABLE statement to wait, but it's not going to do anything about the existence flag.

Reply:
I have an idea though, I will switch the order.  Inform the external files FIRST that the collection is removed, so that they may invalidate.  Then, as the final step, delete the actual table.  If this happens in a transaction it will correctly block any new table creation.